presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedNoAggregationScalarToJoin.java

```diff
@@ -0,0 +1,144 @@
+import static com.facebook.presto.sql.planner.plan.Patterns.lateralJoin;
+import static com.facebook.presto.sql.tree.BooleanLiteral.TRUE_LITERAL;
+
+/**
+ * Scalar filter scan query is something like:
+ * <pre>
+ *     SELECT a,b,c FROM rel WHERE a = correlated1 AND b = correlated2
+ * </pre>
+ * <p>
+ * This optimizer can rewrite to aggregation over a left outer join:
+ * <p>
+ * From:
+ * <pre>
+ * - LateralJoin (with correlation list: [C])
+ *   - (input) plan which produces symbols: [A, B, C]
+ *   - (scalar subquery) Project F
+ *     - Filter(D = C AND E > 5)
+ *       - plan which produces symbols: [D, E, F]
+ * </pre>
+ * to:
+ * <pre>
+ * - Filter(CASE isDistinct WHEN true THEN true ELSE fail('Scalar sub-query has returned multiple rows'))
+ *   - MarkDistinct(isDistinct)
+ *     - LateralJoin (with correlation list: [C])
+ *       - AssignUniqueId(adds symbol U)
+ *         - (input) plan which produces symbols: [A, B, C]
+ *       - non scalar subquery
+ * </pre>
+ * <p>
+ */
+public class TransformCorrelatedNoAggregationScalarToJoin
+        implements Rule<LateralJoinNode>
+{
+    private static final Pattern<LateralJoinNode> PATTERN = lateralJoin()
+            .with(nonEmpty(correlation()));
+
+    @Override
+    public Pattern getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(LateralJoinNode lateralJoinNode, Captures captures, Context context)
+    {
+        PlanNode subquery = context.getLookup().resolve(lateralJoinNode.getSubquery());
+
+        if (!searchFrom(subquery, context.getLookup())
+                .where(EnforceSingleRowNode.class::isInstance)
+                .recurseOnlyWhen(ProjectNode.class::isInstance)
+                .matches()) {
+            return Result.empty();
+        }
+
+        PlanNode rewrittenSubquery = searchFrom(subquery, context.getLookup())
+                .where(EnforceSingleRowNode.class::isInstance)
+                .recurseOnlyWhen(ProjectNode.class::isInstance)
+                .removeFirst();
+
+        LateralJoinNode rewrittenLateralJoinNode = new LateralJoinNode(
+                context.getIdAllocator().getNextId(),
+                new AssignUniqueId(
+                        context.getIdAllocator().getNextId(),
+                        lateralJoinNode.getInput(),
+                        context.getSymbolAllocator().newSymbol("unique", BigintType.BIGINT)),
+                rewrittenSubquery,
+                lateralJoinNode.getCorrelation(),
+                lateralJoinNode.getType(),
+                lateralJoinNode.getOriginSubquery());
+
+        Symbol isDistinct = context.getSymbolAllocator().newSymbol("is_distinct", BooleanType.BOOLEAN);
+        MarkDistinctNode markDistinctNode = new MarkDistinctNode(
+                context.getIdAllocator().getNextId(),
+                rewrittenLateralJoinNode,
+                isDistinct,
+                rewrittenLateralJoinNode.getInput().getOutputSymbols(),
```

Comment:
just use `AssignUniqueId` symbol as distinct symbol (should be faster, less memory)